directives/animate/animate.js

```diff
@@ -0,0 +1,250 @@
+
+	constructor(propertyPart) {
+
+		if (!(propertyPart instanceof PropertyPart) || propertyPart.committer.name !== 'animate') {
+			throw new Error('animation directives must be used with "animation" property');
+		}
+
+		this.id = 0;
+		this.clone = null;
+		this.elem = propertyPart.committer.element;
+		this.state = 'unknown';
+
+	}
+
+	async animate(info) {
+
+		if (this.clone === null) {
+			this.clone = document.createElement('div');
+		}
+		this.elem.parentNode.insertBefore(this.clone, this.elem);
+
+		const id = ++this.id;
+
+		let outerResolve;
+		const onTransitionEnd = (event) => {
+			if (event.target === this.clone) {
+				this.clone.removeEventListener('transitionend', onTransitionEnd);
+				if (this.id === id) {
+					this.clone.remove();
+					this.clone = null;
+					info.onTransitionEnd();
+					['height', 'left', 'opacity', 'position', 'top', 'transform', 'transition', 'width']
+						.forEach((prop) => this.elem.style.removeProperty(prop));
+					outerResolve();
+				}
+			}
+		};
+		this.clone.addEventListener('transitionend', onTransitionEnd);
+
+		Object.assign(this.clone.style, info.clone.start);
+		Object.assign(this.elem.style, info.elem.start);
+
+		if (info.elem.transition) {
+			await new Promise((r) => requestAnimationFrame(r));
+			Object.assign(this.elem.style, info.elem.transition);
+		}
+
+		await new Promise((r) => requestAnimationFrame(r));
+		Object.assign(this.clone.style, info.clone.end);
+		Object.assign(this.elem.style, info.elem.end);
+
+		return new Promise((resolve) => outerResolve = resolve);
+
+	}
+
+	async getElemInfo() {
+
+		await new Promise((r) => requestAnimationFrame(r));
+		const style = window.getComputedStyle(this.elem);
+
+		if (style.display !== 'flex') {
+			this.elem.style.display = 'grid'; // forces margins to collapse during size calculation
+		}
+
+		const rect = this.elem.getBoundingClientRect();
+
+		let top = 0;
+		let left = 0;
+		const offsetParent = this.elem.offsetParent;
+		let e = this.elem;
+		while (e && e !== document && !(e instanceof DocumentFragment)) {
+			if (e === offsetParent) {
+				break;
+			}
+			top += e.offsetTop - (e.scrollTop || 0);
+			left += e.offsetLeft - (e.scrollLeft || 0);
+			e = getComposedParent(e);
+		}
+
+		this.elem.style.removeProperty('display');
+
+		const paddingH = (parseInt(style.paddingLeft) || 0) + (parseInt(style.borderLeftWidth) || 0) + (parseInt(style.paddingRight) || 0) + (parseInt(style.borderRightWidth) || 0);
+		const paddingV = (parseInt(style.paddingTop) || 0) + (parseInt(style.borderTopWidth) || 0) + (parseInt(style.paddingBottom) || 0) + (parseInt(style.borderBottomWidth) || 0);
+		const marginsH = (parseInt(style.marginLeft) || 0) + (parseInt(style.marginRight) || 0);
+		const marginsV = (parseInt(style.marginTop) || 0) + (parseInt(style.marginBottom) || 0);
+
+		let cloneHeight = 0;
+		if (this.clone !== null) {
+			const cloneRect = this.clone.getBoundingClientRect();
+			cloneHeight = cloneRect.height;
+		}
+
+		return {
```

Comment:
Hmmm, are the dimensions here all good regardless of whether the element's `box-sizing` is `border-box` or `content-box`?  Do we end up accounting for the border and padding twice if it's `border-box`?

Reply:
Good catch -- it's broken. Will fix.

Reply:
Glad you noticed this. I didn't read the [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) documentation clearly enough:

> In the standard box model, this would be equal to the width or height property of the element + padding + border-width. But if box-sizing: border-box is set for the element this would be directly equal to its width or height.

So it will always return the right thing, meaning we don't need to worry about padding or border. I've removed that code.

Reply:
Sweet!  👍